### PR TITLE
Separate non-bundled install; generate against current WP-CLI instance.

### PR DESCRIPTION
**Pull request by gitlost**
See https://github.com/wp-cli/handbook/pull/200#issuecomment-362240087

Adds a separate `bin/install_packages.sh` script to be run before running the `handbook` commands that installs the 4 non-bundled packages in `bin/packages`. The `handbook gen-commands` command then just checks that they're installed.

Uses the current WP-CLI instance to generate the doc against rather than invoking a separate instance and insists that it is run with `WP_CLI_PACKAGES_DIR=bin/packages`, and warns if isn't run with `WP_CLI_CONFIG_PATH=/dev/null`.

Empties the `commands/` and `internal-api/` directories first to allow for removed commands/apis. This causes a change in the currently generated doc as `Utils\wp_clear_object_cache()` was deprecated for v1.5.0 so isn't listed.

Adds `--verbose` option to `gen-all` and `gen-commands` to only output individual command listings if set, to make output simpler.

Changes `update_commands_data()` to take `$full` rather than `$parent` arg to simplify calling, and to _~~just~~_ check `when_invoke` closure use arg is _array or_ callable function (otherwise was getting an exception as can be set _~~2~~ both_ ways in `CommandFactory::create_subcommand()`), and sets `repo_url` in subcommand loop on first subcommand found if not already set, and only sets in `$command_data` array if something to set.

Awards "Hack of the Codebase" :1st_place_medal: to @danielbachhuber for getting closure use clause arg with `ReflectionFunction::getStaticVariables()`. Masterful!

_Edit: undid https://github.com/wp-cli/handbook/pull/207/commits/2d24dffa85d49a214eb4ee6e70e6b939f4efe65f as need to check if array (with class/method) or function as the PR originally did._

### CODE REVIEW

**Comment by schlessera** [comment]
> Awards "Hack of the Codebase" 🥇 to @danielbachhuber for getting closure use clause arg with ReflectionFunction::getStaticVariables(). Masterful!

Speaking about hacks... => https://wordpress.slack.com/archives/C02RP4T41/p1502198773228898
Just be glad you don't have regexes in there now... ;)

File: bin/command.php [hunk 1]
```diff
@@ -18,25 +19,35 @@
...
+	 *
+	 * [--verbose]
+	 * : If set will list command pages as they are generated.
 	 *
 	 * @subcommand gen-all
 	 */
-	public function gen_all() {
+	public function gen_all( $args, $assoc_args ) {
+		// Warn if not invoked with null WP_CLI_CONFIG_PATH.
+		if ( '/dev/null' !== getenv( 'WP_CLI_CONFIG_PATH' ) ) {
```
Comment (schlessera) [inline bin/command.php]: That won't work on Windows. However, not sure if we need to be able to generate the docs on Windows machines...

Comment (schlessera) [inline bin/command.php]: If this is a requirement, why not just `putenv()` it to force it?

File: bin/command.php [hunk 2]
```diff
@@ -109,53 +122,46 @@ public function gen_api_docs() {
...
-		$bundled_cmds = array();
-		foreach( $wp['subcommands'] as $k => $cmd ) {
-			if ( in_array( $cmd['name'], array( 'website', 'api-dump', 'handbook' ) ) ) {
-				unset( $wp['subcommands'][ $k ] );
-				continue;
-			}
-			$bundled_cmds[] = $cmd['name'];
+	public function gen_commands( $args, $assoc_args ) {
+		// Check invoked with packages directory set to `bin/packages'.
+		if ( ! preg_match( '/bin\/packages\/?$/', getenv( 'WP_CLI_PACKAGES_DIR' ) ) ) {
```
Comment (schlessera) [inline bin/command.php]: If this is a requirement, why not just `putenv()` it to force it?

File: bin/command.php (hunk 1 above)
Comment (gitlost) [reply bin/command.php]: I did actually consider that, but these are used during bootstrapping so there's no hook available that early? Plus there's an advantage I think to just being explicit and dumb so it's obvious what is happening for such a seldom used command....

Comment (gitlost) [reply bin/command.php]: I did actually consider that also believe it or not - only a warning but it is lazy...

**Pull request merged**